### ISSUE TITLE
Is there any tutorial about how to create a custom Criterion ?

### ISSUE THREAD

**Issue by AlecWong** [torch/nn#875]
Anyone?
Below is one of the criterion provided by Torch7.
```
local MSECriterion, parent = torch.class('nn.MSECriterion', 'nn.Criterion')

function MSECriterion:__init(sizeAverage)
   parent.__init(self)
   if sizeAverage ~= nil then
     self.sizeAverage = sizeAverage
   else
     self.sizeAverage = true
   end
end

function MSECriterion:updateOutput(input, target)
   self.output_tensor = self.output_tensor or input.new(1)
   input.THNN.MSECriterion_updateOutput(
      input:cdata(),
      target:cdata(),
      self.output_tensor:cdata(),
      self.sizeAverage
   )
   self.output = self.output_tensor[1]
   return self.output
end

function MSECriterion:updateGradInput(input, target)
   input.THNN.MSECriterion_updateGradInput(
      input:cdata(),
      target:cdata(),
      self.gradInput:cdata(),
      self.sizeAverage
   )
   return self.gradInput
end 
```
What I should do is mimic it?  Is there any extra jobs should be done(I hope it can support GPU, so?)?

**Comment by clement-masson** [comment]
I have implemented some custom criteria and I personally know two ways of doing so:

1/ You make a C and CUDA implementation and call them from a .lua file exactly like MSECriterion

2/ You make a "pure lua" criterion, ie you implement the actual computation in lua using the math operation available on torch.Tensor  (https://github.com/torch/torch7/blob/master/doc/maths.md)

For simple criteria that can be expressed using only simple operations, the second option may be the best cause it is simpler and faster to write the criteria and will be directly working with all types of Tensors (cpu and gpu).
For more complex criteria, it might be more computationally efficient to implement directly in C/CUDA.

**Comment by AlecWong** [comment]
@clement-masson  Thanks! Option 2 is preferred. 👍 :)

**Comment by rainforest1** [comment]
Where is the tutorial on torch.ch? I am unable to locate it. Please help

**Comment by fmassa** [comment]
@rainforest1 http://torch.ch/docs/developer-docs.html#_
The C/Cuda parts are a bit outdated though, as the underlying interface has changed.

**Comment by moraval** [comment]
@fmassa I need to implement my custom criterion in CUDA (to speed up the calculation). Do you know where could I find some up-to-date information about the interface? Thanks

**Comment by fmassa** [comment]
@moraval Easiest way is to use ffi.
Check for example this [lua file](https://github.com/szagoruyko/imagine-nn/blob/master/ROIPooling.lua), which defines the lua code for creating a `nn` module (criterions are pretty similar).
Once you have your CUDA kernels written, you need to write a C function which will call the kernel, like in [here](https://github.com/szagoruyko/imagine-nn/blob/master/ROIPooling.cu#L84-L115).
After you compile your C code into a shared library, you can then call it from lua pretty easily.
For that, first you need to add the `ffi` definitions that tells lua what are the signatures of the functions that you want to call. This can be found [here](https://github.com/szagoruyko/imagine-nn/blob/master/ffi.lua#L14-L16). Then, you need to tell lua to load the library, which can be found [here](https://github.com/szagoruyko/imagine-nn/blob/master/ffi.lua#L25).

This `ffi` file returns a table with the C functions that were imported. In this library (imagine-nn), they are added to a [C field in the library table](https://github.com/szagoruyko/imagine-nn/blob/master/init.lua#L3), but that is not necessary. Once you have that, you can call your function from lua as done in [here](https://github.com/szagoruyko/imagine-nn/blob/master/ROIPooling.lua#L26-L28).

For a guide on how to use lua ffi, check those tutorials
https://github.com/Atcold/torch-Developer-Guide/tree/master/FFI-so
https://github.com/Atcold/torch-Developer-Guide/tree/master/FFI-Tensor

Hope it helps

**Comment by moraval** [comment]
@fmassa  Thanks a lot! I'll look into it...

**Comment by alex3s** [comment]
@clement-masson , sounds great!

I want to implement Pearsons R2 correlation coefficient criterion using method#2 you mentioned. 
Can you point me to some working examples of method#2 ?

> I have implemented some custom criteria and I personally know two ways of doing so:
> ..
> 2/ You make a "pure lua" criterion, ie you implement the actual computation in lua using the math operation

**Comment by clement-masson** [comment]
@alex3s 
Hi.
You would have to do something like this : 

```
local PearsonsCriterion, parent = torch.class('nn.PearsonsCriterion', 'nn.Criterion')

function PearsonsCriterion:updateOutput(input, target)
   self.output = < compute your criterion here using misc math operation on input and target >
   return self.output
end

function PearsonsCriterion:updateGradInput(input, target)
   self.gradInput:resizeAs(input)
  < compute your the Criterion derivative wrt input here, using misc math operation on input and target >
   return self.gradInput
end
```

**Comment by alex3s** [comment]
@clement-masson 
Thank you. 
This is how it looks so far. Only updateOutput function is correct. Any ideas how to implement the derivative part in updateOutput?
```
local PearsonsCriterion, parent = torch.class('nn.PearsonsCriterion', 'nn.Criterion')

function PearsonsCriterion:__init(weights, sizeAverage)
    parent.__init(self)
    if sizeAverage ~= nil then
        self.sizeAverage = sizeAverage
    else
        self.sizeAverage = true
    end
    if weights ~= nil then
        assert(weights:dim() == 1, "weights input should be 1-D Tensor")
        self.weights = weights
    end
end


function PearsonsCriterion:__len()
    if (self.weights) then
        return #self.weights
    else
        return 0
    end
end


local function pearson(x,y)
    local x = x - x:mean()
    local y = y - y:mean()
    return x:dot(y) / (x:norm() * y:norm())
end


function PearsonsCriterion:updateOutput(input, target)

    assert( input:nElement() == target:nElement(),
    "input and target size mismatch")

    self.output = pearson(input, target) or 0

   return self.output
end

function PearsonsCriterion:updateGradInput(input, target)
   self.gradInput:resizeAs(input)
   --print ({gradInput=self.gradInput})
   
   self.gradInput=(input-target)*math.abs(pearson(input,target))
      
--   local base=pearson(input, target) or 0
--   for i=1, input:size(1) do
--        local new=input:clone()
--        new[i]=new[i]-0.001
--        local newres=pearson(new, target) or 0
--        local diff = 1000*(newres-base)
--        self.gradInput[i]=diff
--   end
   
    -- < compute your the Criterion derivative wrt input here, using misc math operation on input and target >
    --http://www.wolframalpha.com/input/?i=Correlation%5B%7B1,2,3,4%7D,%7B2,1,4,3%7D%5D
    --http://www.wolframalpha.com/input/?i=derivative+of+Correlation%5B%7B1,2,3%7D,%7B2,1,x%7D%5D   
    
   return self.gradInput
end

```

**Comment by lpuglia** [comment]
I know this is an old topic but i found really hard to search for a criterion template, so i thought that the easiest way was to implement a clone of MSECriterion using only LUA code:

`local CloneMSECriterion, parent = torch.class('nn.CloneMSECriterion', 'nn.Criterion')

function CloneMSECriterion:__init(alpha)
   parent.__init(self)
   self.alpha = alpha or 1.1
   self.diff = torch.Tensor()
   self.gradInput = {}
end


function CloneMSECriterion:updateOutput(input, target)

   local N = input:storage():size()

   -- copy
   local cinp = input:clone()

   -- difference input - target
   self.output = cinp - target 

   -- power 2
   self.output:pow(2)

   -- average
   self.output = self.output:sum() / N

   return self.output
end

function CloneMSECriterion:updateGradInput(input, target)

   local N = input:storage():size()
   local cinp = input:clone()
   cinp = cinp - target

   self.gradInput = ((2/(N)) * cinp ) 

   return self.gradInput
end`

Can someone confirm that this is correct?

**Comment by alex3s** [comment]
lpuglia, it looks great!
Are you sure however, that the gradients have to be divided by the number of outputs?

>    self.gradInput = ((2/(N)) * cinp )

**Comment by lpuglia** [comment]
if sizeAverage is set to false in MSECriterion, this is exactly what it supposed to do.

**Comment by alex3s** [comment]
Since the `self.gradInput` is a tensor of the same size as the number of elements of input to `updateGradInput`, it should not be divided. Only the `updateOutput` function is doing averaging (adding all the elements, then dividing by their number). Since there is no summing of the numbers in `updateGradInput`, why dividing them?

**Comment by lpuglia** [comment]
While i was reverse engineering the MSECriterion i notice that you obtain same gradients only when you divide by N! Anyway it is just a matter of relative size, so not very important if you don't mind to have bigger number.

**Comment by hacql2004** [comment]
@lpuglia Hi, thanks for sharing your CloneMSECriterion template. I'm a beginner in torch, when I try to compile your code in my own nn packages, I met some trouble in compiling.
I use the following codes to compile your code, and the torch shows that 'nn scm-1 is now built and installed in /opt/torch/install/ (license: BSD)'. 
However, when I try to call this loss function in my train code, it gives error like ' attempt to call field 'CloneMSECriterion' (a nil value)'. It seems the new module fails to add in the torch. Could you help me to figure out the problem? Thanks again.

Here is my compile code:
`cp CloneMSECriterion .lua opt/torch/extra/nn/
cd opt/torch/extra/nn/
luarocks make rocks/nn-scm-1.rockspec`

**Comment by hacql2004** [comment]
This problem has been solved temporarily by add ' require 'nn/CloneMSECriterion' ' before calling the CloneMSECriterion().

**Comment by dmorris0** [comment]
Is is possible to add additional arguments to a Criterion?  In particular I want to add element-wise weights to BCECriterion.  For example instead of this;

`function CloneBCECriterion:updateOutput(input, target)`

I want to do:

`function CloneBCECriterion:updateOutput(input, target, eweights)`

The problem is that when I define it this way, and pass in weights like this: 

`criterion:forward(input, target, myweights)`

they do not get passed in (eweights is nil).  Any idea the proper way to add additional arguments like this?

Thanks!

**Comment by stu17682** [comment]
Hey Daniel


I had a similar problem where I had to explicitly set the sizeAverage to false.  Depending how you are declaring the cost function this might work for you.  For example in my case:


local weights = torch.zeros(nClasses)


-------------------------------------------------------------------------------

-- some code to make a tensor of weights, one element for each class

--------------------------------------------------------------------------------


criterion = nn.ClassNLLCriterion(weights) -- pass in weights

criterion.sizeAverage = false -- without this line weights have no effect


Hope this helps!


Cheers


Stuart


--

Stuart Millar

PhD Deep Learning / Cyber Security Researcher

CSIT / QUB, Belfast

smillar09@qub.ac.uk

https://www.linkedin.com/in/stuartmillar


________________________________

**Comment by dmorris0** [comment]
Hi Stuart,

Thanks for the note.  Actually for me it is not size averaging that matters.  Rather, I want to include additional parameters so that I can weight image pixels individually (and not based on category).  I found a solution which I described here: https://groups.google.com/d/msg/torch7/zeNHJx7glzE/UayHVwxZBgAJ

Best!
Daniel
